Dmitriy_Slavinskiy

ijustwannatry opened the pull request:
first pull requets

ijustwannatry committed:
HomeWork 1

ijustwannatry committed:
HomeWork 1 done

ijustwannatry committed:
HomeWork 1 done

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -1,9 +1,16 @@
-    System.out.println("Hello homework!");
+    for (String s : args) {
+      if ("error".equals(s)) {
+        System.out.println("\u001B[31mAlarm!\u001B[0m\n");
```
Concatenation would look better

NikolaevArtem commented on src/main/java/homework_1/Main.java:
```diff
@@ -1,9 +1,16 @@
+    for (String s : args) {
+      if ("error".equals(s)) {
+        System.out.println("\u001B[31mAlarm!\u001B[0m\n");
+        System.exit(0);
```
Simple "break;" will be enough, System.exit(0) is not recommended

ijustwannatry committed:
Fixed version after the review

ijustwannatry committed:
HomeWork_2
Traffic light - bug fixing ver_1.0
Pyramid Printer - not started
Random Chars Table - not started

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.0
Pyramid Printer - not started
Random Chars Table - not started

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.1
Pyramid Printer - not started
Random Chars Table - not started

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.1
Pyramid Printer - Done.
Random Chars Table - not started

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.1
Pyramid Printer - Done.
Random Chars Table - Done.

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.1
Pyramid Printer - Done.
Random Chars Table - Done.
README fixed.

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.1
Pyramid Printer - Done.
Random Chars Table - Done.
README fixed.

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.2
Pyramid Printer - Done.
Random Chars Table - Done.
README fixed.

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.2
Pyramid Printer - Done. Ver_1.1
Random Chars Table - Done. Ver_1.0
README fixed.

ijustwannatry committed:
HomeWork_2
Traffic light - Done. ver_2.2
Pyramid Printer - Done. Ver_1.1
Random Chars Table - Done. Ver_1.0
README fixed.

NikolaevArtem commented on src/main/java/homework_2/PyramidPrinter.java:
```diff
@@ -0,0 +1,23 @@
+            try {
+                int levels = Integer.parseInt(reader.readLine());
+                for (int i = 1; i <= levels; i++) {
+                    System.out.println("x".repeat(i));
```
Does not comile. Did I tell everybody to use java 8?

ijustwannatry replied:
I've used another computer for this homework, Didn't expect that the java version settings don't pull from a git branch.

NikolaevArtem commented on src/main/java/homework_2/RandomCharsTable.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.Random;
+import java.util.Scanner;
+
+public class RandomCharsTable {
```
Bug 
![image](https://user-images.githubusercontent.com/58872509/126037235-235fc8a7-6ee3-463e-bf22-6979f729bd94.png)

NikolaevArtem replied:
![image](https://user-images.githubusercontent.com/58872509/126037239-2fb731f1-48fd-4379-b4ac-3fcee13ac3cd.png)

NikolaevArtem replied:
![image](https://user-images.githubusercontent.com/58872509/126037254-4ffa655b-1e3c-4736-bd97-d061de8277aa.png)

NikolaevArtem commented on src/main/java/homework_2/TrafficLight.java:
```diff
@@ -0,0 +1,73 @@
+import java.time.LocalTime;
+import java.time.format.DateTimeParseException;
+
+public class TrafficLight {
```
Works fine

NikolaevArtem commented on src/main/java/homework_2/TrafficLight.java:
```diff
@@ -0,0 +1,73 @@
+        LocalTime time = LocalTime.ofSecondOfDay(seconds);
+        seconds %= 60;
+        if (seconds < 35) {
+            System.out.println("\u001B[32m" + "Green light." + "\u001B[0m" + " Time: " + time);
```
Opt: better move escape strings to constants

NikolaevArtem commented on src/main/java/homework_2/TrafficLight.java:
```diff
@@ -0,0 +1,73 @@
+
+    private static boolean isLocalTime(String str) {
+        try {
+            LocalTime.parse(str);
```
Opt: if null is passes, then exception is not handled

ijustwannatry replied:
I added handling NPE but if I use CTRL + D hotkeys I get an infinite loop. Should I just stop the program if a null is passed?

ijustwannatry replied:
I decided to stop the program. Because if we pass a null from a keyboard, it means we close the input stream and we will not able to pass something else since that moment, isn't it?

NikolaevArtem commented on src/main/java/homework_2/RandomCharsTable.java:
```diff
@@ -0,0 +1,42 @@
+
+public class RandomCharsTable {
+
+    public static void main(String[] args) {
```
It looks difficult to understand. Please simplify

ijustwannatry committed:
HomeWork_2
Fixed after review 1.0

ijustwannatry committed:
HomeWork_2
Fixed after review 1.0

ijustwannatry committed:
HomeWork 2

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
CheckPoint

ijustwannatry committed:
Merge remote-tracking branch 'origin/master' into feature/DmitriySlavinskiy

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

ijustwannatry committed:
checkpoint

NikolaevArtem commented on src/main/java/homework_2/pyramid_printer/PyramidPrinter.java:
```diff
@@ -0,0 +1,117 @@
+    private static String brick = "x";
+    private static String mode = "default";
+
+    PyramidPrinter(String... args) {
```
Having constructor not public is a pretty bad idea

NikolaevArtem commented on src/main/java/homework_2/random_chars_table/RandomCharsTable.java:
```diff
@@ -0,0 +1,81 @@
+import java.util.Random;
+import java.util.Scanner;
+
+public class RandomCharsTable {
```
Most of my tests fail when run together, but pass if run independently. Most probably, it is because you use a lot of static. Make objects where possible.

NikolaevArtem commented:
There are some critical bugs. Please fix

ijustwannatry committed:
After Review

ijustwannatry committed:
After Review

ijustwannatry committed:
After Review

ijustwannatry committed:
HW_2 Done.
Readme refreshed

ijustwannatry committed:
Cosmetic changes.

NikolaevArtem commented on build.gradle:
```diff
@@ -10,6 +10,8 @@ repositories {
 }
 
 dependencies {
+    implementation 'junit:junit:4.13.1'
```
Opt: why do you need this?

ijustwannatry replied:
I've been trying something and IDEa has added this dependency automatically, I think. Removed it.

ijustwannatry committed:
After Review

ijustwannatry committed:
homework_4
CustomFileReader done

ijustwannatry committed:
homework_4
CustomFileReader v1.0.1

ijustwannatry committed:
homework_4
CustomFileReader v1.0.2

ijustwannatry committed:
homework_4

CustomFileReader + tests - Done

ijustwannatry committed:
homework_4
all done

ijustwannatry committed:
homework_4

ijustwannatry committed:
homework_4
all done v1.1

ijustwannatry committed:
homework_4
all done v1.1

ijustwannatry committed:
homework_4
all done v1.1

ijustwannatry committed:
homework_4
all done v1.1

filipchikkirill commented on src/main/java/homework_4/custom_file_reader/CustomFileReader.java:
```diff
@@ -0,0 +1,181 @@
+            this.filePath = PATH + fileName;
+        }
+        file = new File(filePath);
+    }
```
Why not use what you have already written? --->
```
public CustomFileReader(String fileName) {
    this(PATH, fileName);
}
```

ijustwannatry committed:
homework_4
all done v1.2

ijustwannatry committed:
homework_5

ijustwannatry committed:
homework_5

ijustwannatry committed:
homework_5

ijustwannatry committed:
homework_5

ijustwannatry committed:
homework_5
readme fixed

ijustwannatry committed:
homework_6
readme

ijustwannatry committed:
homework_6

ijustwannatry committed:
homework_6

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
homework_7

ijustwannatry committed:
Merge remote-tracking branch 'origin/feature/DmitriySlavinskiy' into feature/DmitriySlavinskiy

ijustwannatry committed:
homework_7

ijustwannatry committed:
HW_7 edited

ijustwannatry committed:
HW_7 edited

ijustwannatry committed:
Classes added
work with input

ijustwannatry committed:
In process

ijustwannatry committed:
homeworks fixed

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
Tests is needed for SeaBattle

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
SeaBattle in process

ijustwannatry committed:
SeaBattle Done

ijustwannatry committed:
SeaBattle Done

ijustwannatry committed:
Update README.md

NikolaevArtem commented on src/main/java/course_project/Main.java:
```diff
@@ -0,0 +1,8 @@
+public class Main {
+
+    public static void main(String[] args) {
+        new SeaBattle().run();
```
When the opponent hits, I don't see him shooting again. So doesn't matter does he hit or miss, next turn is mine

NikolaevArtem replied:
Manual ship placement is comfortable, also automatic is good

NikolaevArtem replied:
Nice colour, testing is visual

NikolaevArtem replied:
Well done! One bug to fix, except that - looks good! I like the nice visual representation, quite comfortable interacting with the console, cool I can choose automatic ship placement. Also, good abstractions, where needed. And good knowledge of Java Core functionality and Java libraries. 

To make better: 
Abstractions are mostly separated by function, but usually, they are separated first by structure (model, service, printer, util, dao, etc). And so models don't usually have logic. 
Better use methods with return values, void methods are not thread-safe and difficult to test

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,83 @@
+
+
+
+public final class ImmutableClass {
```
Actually, your class is mutable: 
![image](https://user-images.githubusercontent.com/58872509/134018686-f9797622-1548-4608-8c35-cf6fdd22b72b.png)

ijustwannatry replied:
Fixed.

NikolaevArtem commented on src/main/java/homework_4/custom_annotation/Info.java:
```diff
@@ -0,0 +1,9 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.FIELD)
+@Inherited
+public @interface Info {
```
Opt: please add examples of usage if this annotation in Main method. I can't actually see how it works with console. As far as I understand it prints fields when printInfo is called, right?

ijustwannatry replied:
I've made a test for showing how it works. The last test in the ChairConstructorAnnotationTest. I use @Info for a new Class Cat, that doesn't have any relationship to the "chair classes". It allows to prints only marked fields.

NikolaevArtem commented on src/main/java/homework_4/custom_annotation/Chair.java:
```diff
@@ -0,0 +1,27 @@
+import java.time.DayOfWeek;
+
+@ChairConstructorAnnotation(dayOfProducing = DayOfWeek.MONDAY)
+public abstract class Chair {
```
Good, some complicated logic with annotations!

NikolaevArtem commented on src/main/java/homework_4/custom_file_reader/CustomFileReader.java:
```diff
@@ -0,0 +1,181 @@
+import java.util.Scanner;
+import java.util.stream.Collectors;
+
+public class CustomFileReader {
```
Good, many different cases!

NikolaevArtem commented on src/main/java/homework_4/singleton/Main.java:
```diff
@@ -0,0 +1,7 @@
+package homework_4.singleton;
+
+public class Main {
+    public static void main(String[] args) {
+
+    }
+}
```
Opt: why do you need it?

ijustwannatry replied:
Forgot to delete it after checks... Fixed.

NikolaevArtem commented on src/main/java/homework_4/singleton/Singleton.java:
```diff
@@ -0,0 +1,28 @@
+package homework_4.singleton;
+
+public class Singleton {
```
Great job, thread-safe singletone!

NikolaevArtem commented on src/main/java/homework_4/singleton/Singleton.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public static Singleton getInstance(String nameOfInstance) {
+        Singleton local = INSTANCE;
+        if (local == null) {
+            synchronized (Singleton.class) {
+                local = INSTANCE;
+                if (local == null) {
+                    INSTANCE = local = new Singleton(nameOfInstance);
+                }
+            }
+        }
+        return local;
```
Opt: it seems there are some unnecessary branching and code conventions are not followed

NikolaevArtem replied:
I see you used Double Checked Locking, then I hope you really understood what you're doing and not just copypasted it)

NikolaevArtem commented on src/main/java/homework_5/custom_regex_matcher/CustomRegexMatcher.java:
```diff
@@ -0,0 +1,19 @@
+
+import homework_2.ConsoleSafeReader;
+
+public class CustomRegexMatcher {
```
Good, thorough testing

NikolaevArtem commented on src/main/java/homework_6/map_problems_generator/MapProblemsCollisionGenerator.java:
```diff
@@ -0,0 +1,34 @@
+import java.util.HashMap;
+import java.util.Objects;
+
+public class MapProblemsCollisionGenerator {
```
I don't see collisions implementation and example

ijustwannatry replied:
I've made examples in the Main class. Should I make a test class instead?

ijustwannatry replied:
Oh... I got it. Sorry, I'll fix it right now!

NikolaevArtem commented on src/main/java/homework_7/kitten_to_cat_function/Main.java:
```diff
@@ -0,0 +1,13 @@
+package homework_7.kitten_to_cat_function;
+
+public class Main {
```
Opt: i would try to create lamda that also converts types, not just values (as example, favorite toy to favorite color or smth like that)

NikolaevArtem commented on src/test/java/course_project/SeaBattleTest.java:
```diff
@@ -0,0 +1,44 @@
+    ;
+
+    @Test
+    void givenAllCoordinates_whenRun_thenEndGame() {
```
Thank you for informative test cases names

NikolaevArtem reviewed (changes requested):
Good job! Although you need to fix HW_3, HW_6 and course project. Write me once you're done

ijustwannatry commented on src/main/java/course_project/Main.java:
```diff
@@ -0,0 +1,8 @@
+public class Main {
+
+    public static void main(String[] args) {
+        new SeaBattle().run();
```
Thanks. About bug:
![image](https://user-images.githubusercontent.com/43565835/134030301-7f44d3c1-fc02-4c22-860a-123fb1f1bf00.png)
I print all enemy's moves. If he hits a ship, he will move again till a miss
I can print the field after every move, but it overloads output and looks confusing.

ijustwannatry committed:
Bug HW_3 fixed

ijustwannatry commented on src/main/java/homework_4/singleton/Singleton.java:
```diff
@@ -0,0 +1,28 @@
+    }
+
+    public static Singleton getInstance(String nameOfInstance) {
+        Singleton local = INSTANCE;
+        if (local == null) {
+            synchronized (Singleton.class) {
+                local = INSTANCE;
+                if (local == null) {
+                    INSTANCE = local = new Singleton(nameOfInstance);
+                }
+            }
+        }
+        return local;
```
I read about it before using it. I hope I understood :)

ijustwannatry committed:
Bug HW_3 fixed
Mistake from HW_6 fixed

ijustwannatry committed:
Upd: SeaBattle
added pause between moves when the enemy hit a ship.
Fixed the output at a moment where there is a question considered who will make the first move

ijustwannatry committed:
Upd: SeaBattle
the field for a reader now is final

ijustwannatry committed:
Upd: SeaBattle
After destroying a ship, a player continues to shot

NikolaevArtem commented on src/main/java/homework_3/ImmutableClass.java:
```diff
@@ -0,0 +1,84 @@
+        this.password = Arrays.copyOf(password, password.length);
+    }
+
+    public final ImmutableClass getChangedInstance(String className) {
+        return new ImmutableClass(className);
+    }
+
+    public final ImmutableClass getChangedInstance(int yearOfBirth) {
+        return new ImmutableClass(yearOfBirth);
+    }
+
+    public final ImmutableClass getChangedInstance(String className, int yearOfBirth) {
+        return new ImmutableClass(className, yearOfBirth);
+    }
+
+    public final ImmutableClass getChangedInstance(String className, int yearOfBirth, char[] password) {
+        return new ImmutableClass(className, yearOfBirth, password);
+    }
```
You lose data calling first 3 methods, they are just like constructors, not copies. 4th is not needed

NikolaevArtem replied:
Moreover, your class is still mutable, as you pass reference of an inner object to another ImmutableClass object

ijustwannatry replied:
I agree about methods, it's my mistake. Didn't think thoroughly. But about passing references I don't understand. When we pass ints, we copy a value. When I pass String, yes, I give a reference, but the field is final and we can't change the reference to a new object, plus the String is immutable. The only way to change something was in returning the char array, but after a fixing, I can't change anything in my class. I don't understand how this class can be mutable.

ijustwannatry committed:
upd: HW_3
need to think about immutable class structure

ijustwannatry committed:
upd: HW_3

ijustwannatry committed:
upd: HW_3

ijustwannatry committed:
upd: HW_6

ijustwannatry committed:
upd: HW_6

NikolaevArtem reviewed (approved):
HW approved!
